### DISCUSSION TITLE
Deprecation thrown as exception

### DISCUSSION THREAD

**Discussion by arsrobota** [symfony/symfony#52706] [Q&A]
Hello,
I really stumbled upon an error while testing my code after a major upgrade from 5.4 to 6.3. Executing the console command I defined, it fails and displays : 
`13:44:55 ERROR     [deprecation] User Deprecated: AbstractPlatform::canEmulateSchemas() is deprecated. (AbstractPlatform.php:4132 called by DefaultQuoteStrategy.php:45, https://github.com/doctrine/dbal/pull/4805, package doctrine/dbal)
[
  "exception" => ErrorException {
    #message: "User Deprecated: AbstractPlatform::canEmulateSchemas() is deprecated. (AbstractPlatform.php:4132 called by DefaultQuoteStrategy.php:45, https://github.com/doctrine/dbal/pull/4805, package doctrine/dbal)"
    #code: 0
    #file: "C:\Users\NVYT6223\projets\symfony\leclair\leclairgp_8.2\vendor\doctrine\deprecations\lib\Doctrine\Deprecations\Deprecation.php"
    #line: 210
    #severity: E_USER_DEPRECATED
    ...
  }
]`
I first [asked on doctrine/dbal GitHub](https://github.com/doctrine/dbal/issues/6228) and the answer (thanks to derrabus) revealed an error handling misconfiguration. But I can hardly figure out what configuration I need to change (PHP ? Symfony ?). I checked my php.ini : 
`
error_reporting = E_ALL
display_errors = On
`
and my Symfony framework.yaml : 
`
    php_errors:
        log: true
`

What can I concretely do to keep deprecations logged, but not throw them as exceptions ?

Cheers,

Alecz

**Comment by ghost** [comment]
in what env you get this exceptions? 
Do you use Symfony runtime inside index.php or the previous content of this file? If the latter then please provide your index.php.
Make sure your application uses ErrorHandler class (e.g. you can die() inside handleError and handleException)

**Reply by arsrobota** [reply]
It's in dev environment. My application calls handleError but not handleException.
`Do you use Symfony runtime inside index.php or the previous content of this file?` I'm not quite sure to understand your question, here is index.php
`<?php

use App\Kernel;

require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

return function (array $context) {
    return new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
};
`

**Reply by ghost** [reply]
Ok, you use symfony runtime so it's responsible for registering error handler.

The problem is that something is turning deprecations into errors:
`13:44:55 ERROR     [deprecation] `

Can you check if this line executed:
```php
        if ($throw) {
            throw $errorAsException;
        }
```
and also tell us what is the value of ErrorHandler::$thrownErrors property.
And please check the value of `debug.error_handler.throw_at` parameter.

**Reply by arsrobota** [reply]
In ErrorHandler `throw $errorAsException;` is not executed. Right after this condition, `$this->throwErrors` is equal to -24577.
About `debug.error_handler.throw_at` I couldn't find it doing a `debug:config framework`. Did Is earch at the wrong place ?

**Reply by ghost** [reply]
`debug.error_handler.throw_at` is a container parameter. You can see its value with: `debug:container --parameter=debug.error_handler.throw_at`

**Reply by ghost** [reply]
24577 (I don't understand why this minus sign is there) is a bitmask for E_ERROR, E_USER_DEPRECATED, E_DEPRECATED.

**Reply by arsrobota** [reply]
OK, I often forget to debug the container... So `debug.error_handler.throw_at`  value is `-1`

**Reply by ghost** [reply]
Ok, -1 is the same as E_ALL.
Something weird happens in your app. Please do some more debugging, maybe use xDebug and stepdebug the script. Try setting the DOCTRINE_DEPRECATIONS env var to `track` (default is `trigger`) - this should fix the case with doctrine deprecations, but won't fix the problem with throwing a deprecation as exception (or turning deprecation into exception).

Please try to trigger a deprecation on your own. Somewhere in you code add `trigger_error('test', \E_USER_DEPRECATED);` and execute thhe code with this line. Does it throw an exception?

**Reply by ghost** [reply]
P.S. Does your web app work? Looking at the error you provided in the description I can see it's coming from the console command. When you send an HTTP request to your app (e.g. open app in the browser) does it work? When you open a profiler and go to the **Logs** tab, can you see the same log right there?

**Reply by arsrobota** [reply]
Ok, thanks for your suggestions, I'll try as soon as possible. Actually it is a console app, I cannot try with a brosser, but the command does not break with the error; but nothing happens, it seems to freezer.

**Reply by ghost** [reply]
The discussion's title says **"Deprecation thrown as exception"** and in the description you write: **"Executing the console command I defined, it fails and displays :"**
Now you say **"but the command does not break with the error; but nothing happens, it seems to freezer"**

Don't you think this is a big difference? If you want to get free help, start describing your problems appropriately. Have a good weekend.

**Reply by arsrobota** [reply]
Thank you very much, it is always a pleasure to talk with you
